src/descriptor/mod.rs

```diff
@@ -68,6 +75,258 @@ pub enum Descriptor<Pk: MiniscriptKey> {
+            "No key found after origin description",
+        ))?;
+        let xpub = bip32::ExtendedPubKey::from_str(xpub_str)
+            .map_err(|_| DescriptorKeyParseError("Error while parsing xpub."))?;
+
+        let mut is_wildcard = false;
+        let derivation_path = key_deriv
+            .filter_map(|p| {
+                if !is_wildcard && p == "*" {
+                    is_wildcard = true;
+                    None
+                } else if is_wildcard {
+                    Some(Err(DescriptorKeyParseError(
+                        "'*' may only appear as last element in a derivation path.",
+                    )))
+                } else {
+                    Some(bip32::ChildNumber::from_str(p).map_err(|_| {
+                        DescriptorKeyParseError("Error while parsing key derivation path")
+                    }))
+                }
+            })
+            .collect::<Result<bip32::DerivationPath, _>>()?;
+
+        if (&derivation_path).into_iter().all(|c| c.is_normal()) {
+            Ok((xpub, derivation_path, is_wildcard))
+        } else {
+            Err(DescriptorKeyParseError(
+                "Hardened derivation is currently not supported.",
+            ))
+        }
+    }
+
+    /// Derives a new key using the path if self is a wildcard xpub. Otehrwise returns a copy of
+    /// self.
+    ///
+    /// Panics if derivation path contains a hardened child number
+    pub fn derive(&self, path: &[bip32::ChildNumber]) -> DescriptorPublicKey {
+        debug_assert!(path.into_iter().all(|c| c.is_normal()));
+
+        match *self {
+            DescriptorPublicKey::ShortPub(ref pk) => DescriptorPublicKey::ShortPub(pk.clone()),
+            DescriptorPublicKey::XPub(ref xpub) => {
+                if xpub.is_wildcard {
+                    DescriptorPublicKey::XPub(DescriptorXPub {
+                        origin: xpub.origin.clone(),
+                        xpub: xpub.xpub.clone(),
+                        derivation_path: (&xpub.derivation_path)
+                            .into_iter()
+                            .chain(path.iter())
+                            .cloned()
+                            .collect(),
+                        is_wildcard: false,
+                    })
+                } else {
+                    self.clone()
+                }
+            }
+        }
+    }
+}
+
+impl MiniscriptKey for DescriptorPublicKey {
+    type Hash = hash160::Hash;
+
+    fn to_pubkeyhash(&self) -> Self::Hash {
+        match *self {
+            DescriptorPublicKey::ShortPub(ref spub) => spub.key.to_pubkeyhash(),
+            DescriptorPublicKey::XPub(ref xpub) => {
+                let ctx = secp256k1::Secp256k1::verification_only();
+                xpub.xpub
+                    .derive_pub(&ctx, &xpub.derivation_path)
+                    .expect("Won't fail, only normal derivations")
+                    .public_key
+                    .to_pubkeyhash()
```

Comment:
This could be extracted into an implementation of `MiniscriptKey` for `DescriptorXPub`. Also the derivation could be used from `ToPublicKey`.

Reply:
Will give a look

Reply:
> Also the derivation could be used from ToPublicKey.

Oh, definitely. Will swap the two commits then.

> This could be extracted into an implementation of MiniscriptKey for DescriptorXPub.

Do we really want that ? If we do we probably want it for SinglePub too..

Reply:
> Do we really want that ? If we do we probably want it for SinglePub too.

You are right, let's implement it once there is an actual need for it.